assets/scripts/build_gallery.py

```diff
@@ -228,132 +424,67 @@ def build_gallery(
-            )
-            if is_satisfied or req in just_installed:
-                continue
-            logger.debug(f"Executing command: 'pip install \"{req}\"'")
-            status_code = execute_shell_command(f'pip install "{req}"')
-            if status_code == 0:
-                just_installed.add(req)
-            else:
-                expectation_tracebacks.write(
-                    f"\n\n----------------\n{expectation} ({group})\n"
-                )
-                expectation_tracebacks.write(f"Failed to pip install {req}\n\n")
-
-        if group != "core":
-            logger.debug(f"Importing {expectation}")
-            try:
-                if group == "great_expectations_experimental":
-                    importlib.import_module(f"expectations.{expectation}", group)
-                else:
-                    importlib.import_module(f"{group}.expectations")
-            except (ModuleNotFoundError, ImportError, Exception) as e:
-                logger.error(f"Failed to load expectation: {expectation}")
-                print(traceback.format_exc())
-                expectation_tracebacks.write(
-                    f"\n\n----------------\n{expectation} ({group})\n"
-                )
-                expectation_tracebacks.write(traceback.format_exc())
-                failed_to_import_set.add(expectation)
-                continue
+    # Install any requirements needed by selected contrib Expectations
+    contrib_requirements_set = set()
+    for _info in expectations_info.values():
+        contrib_requirements_set.update(_info["requirements"])
+    installed = install_necessary_requirements(list(contrib_requirements_set))
 
+    # Get Expectation instances and run diagnostics
+    expectation_instances = get_expectation_instances(expectations_info)
+    for expectation, instance in sorted(expectation_instances.items()):
         logger.debug(f"Running diagnostics for expectation: {expectation}")
         try:
-            impl = great_expectations.expectations.registry.get_expectation_impl(
-                expectation
-            )
-            diagnostics = impl().run_diagnostics(
+            diagnostics = instance.run_diagnostics(
                 ignore_suppress=ignore_suppress,
                 ignore_only_for=ignore_only_for,
                 for_gallery=True,
                 debug_logger=logger,
                 only_consider_these_backends=only_consider_these_backends,
                 context=context,
             )
+            diagnostic_objects[expectation] = diagnostics
             checklist_string = diagnostics.generate_checklist()
             expectation_checklists.write(
-                f"\n\n----------------\n{expectation} ({group})\n"
+                f"\n\n----------------\n{expectation} ({expectations_info[expectation]['package']})\n"
             )
             expectation_checklists.write(f"{checklist_string}\n")
```

Comment:
What do you think about factoring this out (from `diagnostic_objects[expectation] = diagnostics` to the `except:`) into a separate method? `write_expectation_diagnostics()` or something similar? Then could this step and the `run_diagnostics()` step be skipped when only combining all the results from the different backends?

Reply:
How so? Sounds interesting.

Reply:
Might be worth a separate follow-up ticket so it's not lumped in with all these changes.

Reply:
I'm trying to figure out ways to separate this large function up so that we can separate responsibilities better. Something like a `build_gallery_data_for_backend()` and a `merge_gallery_data()`. Pulling out code like `write_expectation_diagnostics()` helps me understand what is actually happening at each stage and makes it easier to make sure the right steps are happening.